ignite/distributed/comp_models/xla.py

```diff
@@ -137,19 +137,37 @@ def spawn(  # type: ignore[override]
             "OR": "or",
         }
 
-        def _do_all_reduce(self, tensor: torch.Tensor, op: str = "SUM") -> torch.Tensor:
+        def _do_all_reduce(
+            self, tensor: torch.Tensor, op: str = "SUM", group: Optional[Union[Any, List[int]]] = None
+        ) -> torch.Tensor:
             if op not in self._reduce_op_map:
                 raise ValueError(f"Unsupported reduction operation: '{op}'")
+            if group is not None and not isinstance(group, list):
+                raise ValueError("group should be list of int or list of list of int")
+            elif group is not None and not all(isinstance(item, list) for item in group):
+                raise ValueError("group should be list of int or list of list of int")
+            elif group is not None and not all(isinstance(item, int) for list_ in group for item in list_):
+                raise ValueError("group should be list of int or list of list of int")
             op = self._reduce_op_map[op]
-            xm.all_reduce(op, [tensor])
+            xm.all_reduce(op, [tensor], groups=group)
             return tensor
 
-        def _do_all_gather(self, tensor: torch.Tensor) -> torch.Tensor:
+        def _do_all_gather(self, tensor: torch.Tensor, group: Optional[Union[Any, List[int]]] = None) -> torch.Tensor:
             # from https://github.com/jysohn23/xla/blob/model-parallel-colab/Gather_Scatter_Broadcast_PyTorch_XLA.ipynb
+            if group is not None and not isinstance(group, list):
+                raise ValueError("group should be list of int or list of list of int")
+            elif group is not None and not all(isinstance(item, int) for item in group):
+                raise ValueError("group should be list of int or list of list of int")
+            elif (
+                group is not None
+                and not all(isinstance(item, list) for item in group)
+                and all(isinstance(item, int) for list_ in group for item in list_)
+            ):
+                raise ValueError("group should be list of int or list of list of int")
```

Comment:
We can refactor this code into a method such that `_do_all_gather` and `_do_all_reduce` uses the same code without duplication.
In terms code itself, I see that the message is copied 3 times -> refactor. The last `elif` condition seems weird as you already checked for similar types... Can you rework this code ?

Reply:
Sure, will do :)